### PR TITLE
[multimodal]Upgrade object detection backend to mmdet 3.0

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
Change MMDetection backend to 3.0.

Changes:
- Internal data structures
- Pipeline logics
- Yolox Configs
- Mosaic/mixup tranform logics
- Multigpu support logics
- Basically everything

Benefits:
- Able to use latest work like DINO, RTMDet, Detic, etc.
- Have better data structure now (DataContainer in 2.0 is not as good), and also aligns better with open vocabulary detection's needs.
- Will be stable and do not need big refactors for a long period.

Future works:
- Benchmarking
- Update Presets
- Update examples and tutorials
- Refine tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by tonyhoo** [comment]
Thanks for updating the version. Can you provide some benchmark results and attach them in the description? Thanks

**Comment by FANGAreNotGnu** [comment]
> Thanks for updating the version. Can you provide some benchmark results and attach them in the description? Thanks

Just completed the basic changes. I'll run a small benchmark after working on a PR in autogluon-bench. And will do a complete benchmarking after upgrading presets to more recent models (RTMDet/DINO).

File: multimodal/src/autogluon/multimodal/models/mmdet_image.py [hunk 1]
```diff
@@ -293,45 +298,39 @@ def image_feature_dim(self):
...
+            inputs=data["inputs"],
+            data_samples=data["data_samples"],
+            mode=mode,
+        )
+
+        if mode == "loss":
+            return rets
+        elif mode == "predict":
+            # for detailed data structure, see https://github.com/open-mmlab/mmdetection/blob/main/mmdet/structures/det_data_sample.py
+            return [{BBOX: ret.pred_instances, LABEL: ret.gt_instances} for ret in rets]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/mmdet_image.py:331]: Why does `predict` mode require groundtruth labels?

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 2]
```diff
@@ -1773,15 +1773,8 @@ def _default_predict(
...
-                    # outputs shape: num_batch, 1(["bbox"]), batch_size, 80, n, 5
-                    # output LABEL if exists for evaluations
-                    outputs = [
-                        {BBOX: bbox, LABEL: ele[LABEL][i]} if LABEL in ele else {BBOX: bbox}
-                        for ele in outputs
-                        for i, bbox in enumerate(ele[BBOX])
-                    ]
-
+                    # Unpack outputs for object detection while using single gpu
+                    outputs = [output for batch_outputs in outputs for output in batch_outputs]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1777]: This list comprehension logic looks problematic.

File: multimodal/src/autogluon/multimodal/utils/environment.py [hunk 3]
```diff
@@ -261,15 +260,13 @@ def check_if_packages_installed(problem_type: str):
...
-                f"Encountered error while importing mmcv: {e}. Try to install mmcv: mim install mmcv-full."
-            )
+            raise ValueError(f"Encountered error while importing mmcv: {e}. Try to install mmcv: mim install mmcv.")
 
         try:
             import mmdet
         except ImportError as e:
             raise ValueError(
-                f'Encountered error while importing mmdet: {e}. Try to install mmdet: pip install "mmdet>=2.28, <3.0.0".'
+                f'Encountered error while importing mmdet: {e}. Try to install mmdet: pip install "mmdet>=3.0.0".'
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/environment.py:269]: There are several places using similar error messages regarding installing mmdet and mmcv. Consider defining one shared error message and reuse it.

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 4]
```diff
@@ -143,7 +143,7 @@ model:
     max_img_num_per_col: 2
 
   mmdet_image:
-    checkpoint_name: "yolov3_mobilenetv2_320_300e_coco"
+    checkpoint_name: "yolov3_mobilenetv2_8xb24-320-300e_coco"
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml:146]: Why do we need this change?

File: CI/docker/full_install_image.sh [hunk 5]
```diff
@@ -9,6 +9,6 @@ python3 -m pip install multimodal/[tests]
 python3 -m pip install timeseries/[all,tests]
 python3 -m pip install autogluon/
 
-mim install mmcv-full --timeout 60
-python3 -m pip install --upgrade "mmdet>=2.28, <3.0.0"
-python3 -m pip install --upgrade "mmocr<1.0"
+mim install mmcv --timeout 60
+python3 -m pip install --upgrade "mmdet>=3.0.0"
+# python3 -m pip install --upgrade "mmocr<1.0"  # not compatible with mmcv 2.0
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply CI/docker/full_install_image.sh:14]: `setup_mmcv` is in .github folder and this one is in CI folder. It feels weird to do `source $(dirname "$0")/../.github/workflow_scripts/setup_mmcv.sh`.

File: multimodal/src/autogluon/multimodal/models/mmdet_image.py (hunk 1 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/models/mmdet_image.py:326]: `model(mode="predict")` is used for both predicting and evaluation. Here `gt_instances` will be None if there is no label available. It's easier to include the ground truth here than merging ground truth with result in lit module / predictor.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 2 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:1777]: This is simply unpacking a list with two dimensions (`<num_batches>` batches of `<batch_size>` samples each).

File: multimodal/src/autogluon/multimodal/data/process_mmlab/process_mmlab_base.py [hunk 6]
```diff
@@ -74,6 +74,9 @@ def __init__(
         requires_column_info
             Whether to require feature column information in dataloader.
         """
+        from ...utils import get_mmlab_error_msgs, try_import_mmlabs
+
+        try_import_mmlabs()
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_mmlab/process_mmlab_base.py:79]: `try_import_mmlabs ` seems to be used in one place. Why do we need to use mmcv and mmdet related imports at the beginning of this file and also `try_import_mmlabs` in the same file?

File: multimodal/src/autogluon/multimodal/data/process_mmlab/process_mmlab_base.py [hunk 7]
```diff
@@ -123,13 +121,15 @@ def collate_fn(self, image_column_names: Optional[List] = None, per_gpu_batch_si
...
+        from ...utils import check_if_packages_installed
+
         fn = {}
         if self.requires_column_info:
             assert image_column_names, "Empty image column names."
             for col_name in image_column_names:
                 fn[f"{self.image_column_prefix}_{col_name}"] = StackCollator()
 
-        assert mmcv is not None, "Please install mmcv-full by: mim install mmcv-full."
+        check_if_packages_installed(problem_type=None)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_mmlab/process_mmlab_base.py:132]: `problem_type=None` has no effect. I guess the mmcv check is for the collate function. Do we need to check mmcv here?

**Review by zhiqiangdon — APPROVED** [review]
LGTM. Thanks for the upgrade.

**Pull request merged**